packages/server-wallet/src/onchain-service/index.ts

```diff
@@ -0,0 +1,228 @@
+import {ContractArtifacts} from '@statechannels/nitro-protocol';
+import {providers, Wallet, Contract, BigNumber} from 'ethers';
+import PriorityQueue from 'p-queue';
+
+import {Wallet as ChannelWallet} from '../wallet';
+import {Address, Bytes32} from '../type-aliases';
+
+const BAD_NONCE = "the tx doesn't have the correct nonce";
+const INVALID_NONCE = 'Invalid nonce';
+const NO_TX_HASH = 'no transaction hash found in tx response';
+const UNDERPRICED_REPLACEMENT = 'replacement transaction underpriced';
+const DEFAULT_MAX_RETRIES = 5;
+const KNOWN_ERRORS = [BAD_NONCE, NO_TX_HASH, UNDERPRICED_REPLACEMENT, INVALID_NONCE];
+
+type OnchainServiceEnvironment = {
+  provider: string | providers.JsonRpcProvider;
+  wallet: string | Wallet; // Private key or wallet to send tx
+  transactionRetries?: number; // Maximum number of times a tx is retried
+};
+
+// This is used instead of the ethers `Transaction` because that type
+// requires the nonce and chain ID to be specified, when sometimes those
+// arguments are not known at the time of creating a transaction.
+type MinimalTransaction = {
+  to: Address;
+  value: string;
+  data: string;
+};
+
+type AssetHolderInformation = {
+  assetId: Address;
+  contractAddress: Address;
+};
+
+export interface OnchainServiceInterface {
+  // NOTE: will return void and throw an error in final implementation
+  registerChannel(
+    channelId: Bytes32,
+    assetHolders: AssetHolderInformation[]
+  ): Promise<string | undefined>;
+  submitTransaction(
+    channelId: Bytes32,
+    tx: MinimalTransaction
+  ): Promise<providers.TransactionResponse>;
+
+  // TODO: remove in v1
+  attachChannelWallet(wallet: ChannelWallet): void;
+}
+
+export class OnchainService implements OnchainServiceInterface {
+  private provider: providers.JsonRpcProvider;
+  private wallet: Wallet;
+  private retries: number;
+
+  // NOTE: V0 has an in-memory store, which will eventually be
+  // shifted to more permanent storage
+  // Holds latest channel event by type (i.e. { "fundsDeposited": {...} })
+  private store: Map<string, object> = new Map();
+
+  // Stores references to all contracts in memory
+  private assetHolders: Map<string, AssetHolderInformation & {contract: Contract}> = new Map();
```

Comment:
Per my comment above, if the key is the asset holder address, then you only need to store the contract -- the other asset holder information is already on the contract.

```suggestion
  private assetHolders: Map<string, Contract> = new Map();
```

Reply:
I don't want to continually read the token address from the chain, so I'll keep the mapping type as is, but the input types have changed!

Reply:
ah fair, that sounds good